src/cdi/CPCommon.hh

```diff
@@ -0,0 +1,46 @@
+//----------------------------------*-C++-*----------------------------------//
+/*!
+ * \file   cdi/CPCommon.hh
+ * \author Kendra Long
+ * \date   Mon Jan 19 13:41:01 2001
+ * \brief  Datatypes needed in CPEloss classes
+ * \note   Copyright (C) 2016-2019 Los Alamos National Securty, LLC.
+ *         All rights reserved. */
+//---------------------------------------------------------------------------//
+
+#ifndef __cdi_CPCommon_hh__
+#define __cdi_CPCommon_hh__
+
+namespace rtt_cdi {
+
+namespace constants {
+//! Number of models contained in rtt_cdi::CPModel.
+unsigned int const num_CPModels(2);
+
+//! Number of charged particle types in ParticleType above.
+unsigned int const num_CPs(6);
+} // namespace constants
+
+enum CPModel {
+  ELOSS = 0,        /*!< energy loss model */
+  LA_SCATTERING = 1 /*!< large-angle scattering model */
+};
+
+enum CPModelType {
+  UNASSIGNED_ETYPE =
+      0, /*!< unassigned type; used as a placeholder before deciding type */
+  ANALYTIC_ETYPE = 1, /*!< an Analytic eloss model */
+};
+
+enum CParticleType {
+  UNKNOWN_PART = 0, /*!< unknown particle :-O */
+  ELECTRON = 1,     /*!< rest are pretty self-explanatory... */
+  HYDROGEN = 2,
+  DEUTERIUM = 3,
+  TRITIUM = 4,
+  ALPHA = 5
+};
```

Comment:
Would it be reasonable to use integer ZAIDs (with electrons = -1) in the code instead of this enum? This might lead to a bit less typing when adding particle species, and would match the usage of ZAIDs in CPT data, although things like particle mass will still have to be hardcoded behind branching statements.

Reply:
This `enum` list might become very large otherwise!  While an `int` might suffice for now, we should also consider a heavier custom class or struct.  Doing so would give us a little more flexibility in defining special types like electrons.  The custom class could have ctor and getters that take/give an integer ZAID value so it could be used like an `int`

Reply:
Is there a reason why you are using branching as as opposed to tag dispatching? You could have tags with fixed particles properties built in:

```c++
struct Alpha {
  static constexpr auto mass(){ return 6.644657230E-27 * kilo(grams); }
  static constexpr auto charge(){ return 3.2043532416E-19 * coulomb; }
};
```
Then rather than check a condition everytime you fall through a function, you just dispatch to the appropriate function based on the particle tag. For example, this:

```c++
auto foo(Particle p){
  // setup
  if(particle.type() == CParticleType::ELECTRON){
    // implementation electron
  } else if (particle.type() == CParticleType::ALPHA) {
  // implementation for alpha
 }...
}
```
can then become:

```c++
void do_something_impl(Particle p, Electron){
  // do some stuff
}
void do_something_impl(Particle p, Alpha){
  // do some stuff
}
void do_something(Particle p){
  do_something_impl(p, Particle::Type{});
}

auto foo(Particle p){
  // setup
  do_something(p);
}
```
The benefit of doing it this way is that you do not branch at runtime because the compiler will instantiate a function for the different particles types and call it accordingly. Furthermore, this approach follows the open/close principle in that you can accomodate new particles by simply adding a new function to the overload set as opposed to modify an existing function. Happy to discuss this in person if it is unfamiliar.

Reply:
Maybe even simpler, see https://cppinsights.io/s/38e452e1

Reply:
Why not use an enum class, where the names are local to the enum and their values do not implicitly convert to other types, and without assignment?
enum class CPType
{
  CPElectron,
  CPAlpha,
  CPNone
}

Reply:
Thank you all for the input! I understand what you're saying with regard to the design of the class. I admittedly copy-pasted this code from an existing CDI example to get something I could build against. I will address the simpler comments (formatting, etc.) and leave the WIP in place until we hash this out :)